spring-ai-core/src/main/java/org/springframework/ai/chat/client/ChatClient.java

```diff
@@ -455,48 +577,65 @@ public <T> T entity(Class<T> type) {
+
+				Map<String, Object> context = new ConcurrentHashMap<>();
+				context.putAll(inputRequest.advisorParams);
+				ChatClientRequest advisedRequest = adviseOnRequest(inputRequest, context);
 
 				var processedUserText = StringUtils.hasText(formatParam)
-						? this.request.userText + System.lineSeparator() + "{format}" : this.request.userText;
+						? advisedRequest.userText + System.lineSeparator() + "{spring.ai.soc.format}"
+						: advisedRequest.userText;
 
-				Map<String, Object> userParams = new HashMap<>(this.request.userParams);
+				Map<String, Object> userParams = new HashMap<>(advisedRequest.userParams);
 				if (StringUtils.hasText(formatParam)) {
-					userParams.put("format", formatParam);
+					userParams.put("spring.ai.soc.format", formatParam);
 				}
 
-				var messages = new ArrayList<Message>(this.request.messages);
+				var messages = new ArrayList<Message>(advisedRequest.messages);
 				var textsAreValid = (StringUtils.hasText(processedUserText)
-						|| StringUtils.hasText(this.request.systemText));
+						|| StringUtils.hasText(advisedRequest.systemText));
 				if (textsAreValid) {
-					if (StringUtils.hasText(this.request.systemText) || !this.request.systemParams.isEmpty()) {
+					if (StringUtils.hasText(advisedRequest.systemText) || !advisedRequest.systemParams.isEmpty()) {
 						var systemMessage = new SystemMessage(
-								new PromptTemplate(this.request.systemText, this.request.systemParams).render());
+								new PromptTemplate(advisedRequest.systemText, advisedRequest.systemParams).render());
 						messages.add(systemMessage);
 					}
 					UserMessage userMessage = null;
 					if (!CollectionUtils.isEmpty(userParams)) {
 						userMessage = new UserMessage(new PromptTemplate(processedUserText, userParams).render(),
-								this.request.media);
+								advisedRequest.media);
 					}
 					else {
-						userMessage = new UserMessage(processedUserText, this.request.media);
+						userMessage = new UserMessage(processedUserText, advisedRequest.media);
 					}
 					messages.add(userMessage);
 				}
-				if (this.request.chatOptions instanceof FunctionCallingOptions functionCallingOptions) {
-					if (!this.request.functionNames.isEmpty()) {
-						functionCallingOptions.setFunctions(new HashSet<>(this.request.functionNames));
+
+				if (advisedRequest.chatOptions instanceof FunctionCallingOptions functionCallingOptions) {
+					if (!advisedRequest.functionNames.isEmpty()) {
+						functionCallingOptions.setFunctions(new HashSet<>(advisedRequest.functionNames));
 					}
-					if (!this.request.functionCallbacks.isEmpty()) {
-						functionCallingOptions.setFunctionCallbacks(this.request.functionCallbacks);
+					if (!advisedRequest.functionCallbacks.isEmpty()) {
+						functionCallingOptions.setFunctionCallbacks(advisedRequest.functionCallbacks);
+					}
+				}
+				var prompt = new Prompt(messages, advisedRequest.chatOptions);
+				var chatResponse = this.chatModel.call(prompt);
```

Comment:
This is where we can add a chain of 'around advice' so that one can help with the reliability of calls to the model, for example to enforce json, add 'the functions available to call are' and other types of validation with actions to take on failure to try and remediate.